Rules/CommonScripts/plants_regrow.as

```diff
@@ -0,0 +1,199 @@
+#define SERVER_ONLY
+
+const bool moss_stone = false; //should stone turn into mossy stone
+
+const f32 grass_grow_chance = 10;
+const f32 bush_grow_chance = 1;
+const f32 moss_stone_chance = 2;
+
+const u32 moss_time_min = 300*30*30; //10 minutes = 300*30*30
+
+u8 random_time = 120;
+
+TileInfo@[] tiles;
+
+class TileInfo
+{
+	Vec2f coords;
+	u32 place_time;
+	Tile tile;
+
+	TileInfo() {};
+
+	TileInfo(Vec2f _coords, u32 _place_time, Tile _tile)
+	{
+		coords = _coords;
+		place_time = _place_time;
+		tile = _tile;
+	};
+
+	bool mossTime()
+	{
+		return (getGameTime() - place_time > moss_time_min);
+	}
+
+	u8 grassLuck()
+	{
+		u8 luck = 0;
+		CMap@ map = getMap();
+
+		Tile tile_left = map.getTile(coords-Vec2f(map.tilesize, map.tilesize));
+		Tile tile_right = map.getTile(coords-Vec2f(-map.tilesize, map.tilesize));
+
+		if (tile_left.type == CMap::tile_grass) luck += 15;
+		if (tile_right.type == CMap::tile_grass) luck += 15;
+
+		return luck;
+	}
+
+	u8 mossLuck()
+	{
+		u8 luck = 0;
+		CMap@ map = getMap();
+
+		Tile tile_left = map.getTile(coords - Vec2f(map.tilesize, 0));
+		Tile tile_right = map.getTile(coords - Vec2f(-map.tilesize, 0));
+		Tile tile_above = map.getTile(coords - Vec2f(0, map.tilesize));
+		Tile tile_below = map.getTile(coords - Vec2f(0, -map.tilesize));
+
+		if (map.isTileGround(tile_below.type) || map.isTileBedrock(tile_below.type) || tile_below.type == CMap::tile_castle_moss || tile_below.type == CMap::tile_castle_back_moss) {luck += 2;}
+		if (map.isTileGround(tile_above.type) || map.isTileBedrock(tile_above.type) || tile_above.type == CMap::tile_castle_moss || tile_above.type == CMap::tile_castle_back_moss) {luck += 2;}
+		if (map.isTileGround(tile_left.type) || map.isTileBedrock(tile_left.type) || tile_left.type == CMap::tile_castle_moss || tile_left.type == CMap::tile_castle_back_moss) {luck += 2;}
+		if (map.isTileGround(tile_right.type) || map.isTileBedrock(tile_right.type) || tile_right.type == CMap::tile_castle_moss || tile_right.type == CMap::tile_castle_back_moss) {luck += 2;}

```

Comment:
There's a better way of writing this without repeating yourself. This also allows you to add the corner tiles without copying and pasting chunks of code if you choose to.

Ideally, someone should be an engine-side method called `isTileMossy` so you don't have to hardcode specific mossy tiles, but this will do for now I guess.

```suggestion
		Vec2f[] adjacentTilePos = {
			Vec2f(map.tilesize, 0),   //left
			Vec2f(-map.tilesize, 0),  //right
			Vec2f(0, map.tilesize),   //up
			Vec2f(0, -map.tilesize)   //down
		};

		for (u32 i = 0; i < adjacentTilePos.size(); i++)
		{
			Tile tile = map.getTile(coords - adjacentTilePos[i]);

			bool ground = map.isTileGround(tile.type);
			bool bedrock = map.isTileBedrock(tile.type);
			bool mossyCastle = tile.type == CMap::tile_castle_moss || tile.type == CMap::tile_castle_back_moss;

			if (ground || bedrock || mossyCastle)
			{
				luck += 2;
			}
		}
```